test/spec/modules/haloIdSystem_spec.js

```diff
@@ -0,0 +1,31 @@
+import { haloIdSubmodule, storage } from 'modules/haloIdSystem.js';
+import { server } from 'test/mocks/xhr.js';
+import * as utils from 'src/utils.js';
+
+describe('HaloIdSystem', function () {
+  describe('getId', function() {
+    it('gets a haloId', function() {
+      const config = {
+        params: {}
+      };
+      const callbackSpy = sinon.spy();
+      const callback = haloIdSubmodule.getId(config).callback;
+      callback(callbackSpy);
+      const request = server.requests[0];
+      expect(request.url).to.eq(`https://id.halo.ad.gt/api/v1/pbhid`);
+      request.respond(200, { 'Content-Type': 'application/json' }, JSON.stringify({ haloId: 'testHaloId1' }));
+      expect(callbackSpy.lastCall.lastArg).to.deep.equal({haloId: 'testHaloId1'});
+    });
+
+    it('gets a cached haloid', function() {
+      const config = {
+        params: {}
+      };
+      storage.setDataInLocalStorage('auHaloId', 'tstCachedHaloId1');
```

Comment:
It is not recommended to actually have the tests themselves setting localStorage / cookies.

These can become flakey depending on which browsers they are run in.

See https://github.com/prebid/Prebid.js/blob/master/test/spec/modules/rubiconAnalyticsAdapter_spec.js#L585-L589 as an example of mocking your storage object.

Then you can mock it's return like so: https://github.com/prebid/Prebid.js/blob/master/test/spec/modules/rubiconAnalyticsAdapter_spec.js#L1250


Also do not forget to call restore() on each stub in the afterEach's => https://github.com/prebid/Prebid.js/blob/master/test/spec/modules/rubiconAnalyticsAdapter_spec.js#L618-L620

Reply:
i hadn't seen this pattern before, thanks for the info.